keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,340 @@
+  </snip>
+}
+```
+
+```go
+type ServiceAccount struct {
+  <snip>
+  // ImageDecryptSecrets is a list of references to secrets in the same namespace to use for decrypting any encrypted images
+  // in pods that reference this ServiceAccount. 
+  // +optional
+  ImageDecryptSecrets []LocalObjectReference `json:"imageDecryptSecrets,omitempty" protobuf:"bytes,5,rep,name=imageDecryptSecrets"`
+  </snip>
+```
+
+An unspecified `nil` or empty `""` ImageDecryptSecret is equivalent to the backwards-compatible
+default behavior as if the ImageDecryptSecret feature is disabled.
+
+#### Examples
+
+Suppose we operate a cluster that lets users create a secret of type `ImageDecryptSecret`. 
+
+For the kubectl command line, we propose that the user create a secret of type `image-decrypt` and give it a name. One or multiple private keys may then be stored under this secret.
+
+```bash
+kubectl create secret image-decrypt <secret name> --decrypt-secret=/path/to/private_key[:<password>] [--decrypt-secret=/path/to/private_key[:<password>]]
+```
+`private key` - Private keys represented in format DER/PEM or GPG private keyrings.  These keys will be used by JWE/PKCS7/PGP protocols.
+
+`password` (optional) - password in *cleartext* if the given private key is protected with it.
+
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --decrypt-secret="/home/user/keys/private.key":"password"
+```
+
+When a user creates a workload, they can choose the desired ImageDecryptSecret to use 
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: localhost:5000/nginx:enc
+    ports:
+    - containerPort: 80
+  imageDecryptSecrets:
+  - name: keysecret
+```
+
+ImageDecryptSecrets can be added to the `service account` by,
+
+```bash
+kubectl patch serviceaccount <account name> -p '{"imageDecryptSecrets":[{"name":<secret name>]}'
+```
+or while creating a secret account,
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  creationTimestamp: 2015-08-07T22:02:39Z
+  name: default
+  namespace: default
+  selfLink: /api/v1/namespaces/default/serviceaccounts/default
+  uid: 052fb0f4-3d50-11e5-b066-42010af0d7b6
+secrets:
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret"}]}'
+```
+
+#### Key Secret Definition
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI by bundling them in `DecryptParam` as a part of `PullImageRequest` (see, [Image Handler](#Image-handler)) and `CreateContainerRequest` (see, [Container Handler](#Container-handler)):
+
+
+```protobuf
+// DecryptParam represents a single private key (and optional password) that can be sent to the CRI via `PullImageRequest` as well as `CreateContainerRequest`
+message DecryptParam {
```

Comment:
Do we need to include the protocol in the api?
How does the container runtime knows how to use this?

Reply:
API doesn't need to explicitly define the key protocol. Kubernetes passes the given key(s) to runtime _as is_. The runtime should have the logic to handle the key protocol. e.g. The library that will be used by `CRI-O` to handle decryption already has functions to infer the key protocol from the key data, [as seen here](https://github.com/containers/ocicrypt/blob/master/utils/utils.go#L72).

 I have added the same explanation in KEP doc [here](https://github.com/kubernetes/enhancements/blob/1313f21885de178d3ad746e1cf434bf106b34968/keps/sig-node/20190517-image-decryption.md#key-Secret-definition).

Reply:
It seems that the runtime needs to try different protocols.

If it could be easily specified here, why we let runtime try indefinitely?

Reply:
The protocols (jwe/pkcs7/pgp) are defined in the encrypted image (via OCI annotations), so the runtime does not need to guess it once it gets an image it needs to decrypt. An RSA private key can be used by multiple protocols (jwe/pkcs7), so i'm not sure about specifying a protocol here.

However... the runtime does need to try to figure out the format of the key to parse it though, i.e. DER, PEM encoding. But this is a pretty simple operation since it just looks at the first few bytes to determine the header. Currently the list of encodings are only a handful, so it's not much work.

Reply:
Thanks for the explanation.

I think it is useful to put this into the KEP, so that it is more clear how the integration will work, what should be put into this field.